File: ios/Sources/DensoScannerPlugin/DensoScanner.swift
import Foundation
import Capacitor
import DENSOScannerSDK

@objc public class DensoScanner: NSObject {
    weak var plugin: DensoScannerPlugin?
    
    func setupScanner(scanner: CommScanner) -> Bool {
        var error: NSError? = nil
        scanner.claim(&error)
        if (error != nil) {
            return false
        }
        
        plugin!.notifyListeners(DensoScannerEvents.OnScannerStatusChanged.rawValue, data: ["status":DensoScannerStatusEvents.SCANNER_STATUS_CLAIMED.rawValue])
        plugin!.scannerConnected = true
        plugin!.commScanner = scanner
        plugin!.rfidScanner = scanner.getRFIDScanner()
        scanner.addStatusListener(plugin!)
        updateConnectMode(scanner: scanner, connectMode: plugin!.connectMode)
        
        return true
    }
    
    func updateConnectMode(scanner: CommScanner, connectMode: String) -> CommScannerBtSettings {
        var error: NSError? = nil
        
        let btSettings = scanner.getBtSettings(&error)
        if(plugin!.connectMode == DensoScannerConnectMode.SLAVE.rawValue && btSettings?.mode != .MODE_SLAVE) {
            btSettings?.mode = .MODE_SLAVE
            scanner.setBtSettings(btSettings, error: &error)
        } else if (plugin!.connectMode == DensoScannerConnectMode.MASTER.rawValue && btSettings?.mode != .MODE_MASTER) {
            btSettings?.mode = .MODE_MASTER
            scanner.setBtSettings(btSettings, error: &error)
        }
        
        return btSettings!
    }
    
    func convertDataToString(_ data: Data) -> String? {
        return data.map { String(format: "%02X", $0) }.joined()
    }
    
    func convertDataTohexString(_ data: Data) -> String? {
        return data.map { String(format: "%02X", $0) }.joined(separator: " ")
    }
    
    func stringToBytes(_ string: String) -> [UInt8]? {
        let length = string.count
        if length & 1 != 0 {
            return nil
        }
        var bytes = [UInt8]()
        bytes.reserveCapacity(length/2)
        var index = string.startIndex
        for _ in 0..<length/2 {
            let nextIndex = string.index(index, offsetBy: 2)
            if let b = UInt8(string[index..<nextIndex], radix: 16) {
                bytes.append(b)
            } else {
                return nil
            }
            index = nextIndex
        }
        return bytes
    }
    
    func receiveCommScannerSettings(settings: RFIDScannerSettings, btSettings: CommScannerBtSettings) -> JSObject {
        
        let triggerMode: String = switch settings.scan.triggerMode {
            case .RFID_TRIGGER_MODE_AUTO_OFF:
                DensoScannerTriggerMode.RFID_TRIGGER_MODE_AUTO_OFF.rawValue
            case .RFID_TRIGGER_MODE_MOMENTARY:
                DensoScannerTriggerMode.RFID_TRIGGER_MODE_MOMENTARY.rawValue
            case .RFID_TRIGGER_MODE_ALTERNATE:
                DensoScannerTriggerMode.RFID_TRIGGER_MODE_ALTERNATE.rawValue
            case .RFID_TRIGGER_MODE_CONTINUOUS1:
                DensoScannerTriggerMode.RFID_TRIGGER_MODE_CONTINUOUS1.rawValue
            case .RFID_TRIGGER_MODE_CONTINUOUS2:
                DensoScannerTriggerMode.RFID_TRIGGER_MODE_CONTINUOUS2.rawValue
            }
        
        let session: Int = switch settings.scan.sessionFlag {
        case SessionFlag.SESSION_FLAG_S0:
            0
        case SessionFlag.SESSION_FLAG_S1:
           1
        case SessionFlag.SESSION_FLAG_S2:
            2
        case SessionFlag.SESSION_FLAG_S3:
            3
        }
        
        let connectMode : String = switch btSettings.mode {
        case .MODE_MASTER : DensoScannerConnectMode.MASTER.rawValue
        case .MODE_SLAVE : DensoScannerConnectMode.SLAVE.rawValue
        case .MODE_AUTO : DensoScannerConnectMode.AUTO.rawValue
        }
        
        
        return [
            "triggerMode": triggerMode,
            "powerLevelRead": Int(settings.scan.powerLevelRead),
            "session": session,
            "polarization": settings.scan.polarization.des(),
            "connectMode": connectMode,
        ]
    }
}

extension Polarization {
    func des() -> String {
        switch self {
        case .POLARIZATION_V:
            return DensoScannerPolarization.POLARIZATION_V.rawValue
        case .POLARIZATION_H:
            return DensoScannerPolarization.POLARIZATION_H.rawValue
        case .POLARIZATION_BOTH:
            return DensoScannerPolarization.POLARIZATION_BOTH.rawValue
        }
    }
}

File: android/src/main/java/jp/rdlabo/capacitor/plugin/densoscanner/DensoScanner.java
package jp.rdlabo.capacitor.plugin.densoscanner;

import android.util.Log;
import com.densowave.scannersdk.Common.CommScanner;
import com.densowave.scannersdk.Dto.CommScannerBtSettings;
import com.densowave.scannersdk.Dto.RFIDScannerSettings;
import com.getcapacitor.JSObject;
import java.util.Objects;

public class DensoScanner {

    public JSObject createSettingsResponse(RFIDScannerSettings settings, CommScannerBtSettings btSettings) {
        JSObject response = new JSObject();
        response.put("triggerMode", TriggerModeMapper.toString(settings.scan.triggerMode));
        response.put("powerLevelRead", settings.scan.powerLevelRead);
        response.put("polarization", DensoScannerPolarizationMapper.toString(settings.scan.polarization));
        Integer session =
            switch (settings.scan.sessionFlag) {
                case S0 -> 0;
                case S1 -> 1;
                case S2 -> 2;
                case S3 -> 3;
            };
        response.put("session", session);
        response.put("connectMode", ConnectModeMapper.toString(btSettings.mode));

        Log.d("denso", response.toString());

        return response;
    }

    public CommScannerBtSettings updateConnectMode(CommScanner connectedCommScanner, String connectMode) {
        CommScannerBtSettings btSet = null;
        try {
            btSet = connectedCommScanner.getBtSettings();

            if (Objects.equals(connectMode, "MASTER") && btSet.mode != CommScannerBtSettings.Mode.MASTER) {
                btSet.mode = CommScannerBtSettings.Mode.SLAVE;
                connectedCommScanner.setBtSettings(btSet);
            } else if (Objects.equals(connectMode, "SLAVE") && btSet.mode != CommScannerBtSettings.Mode.SLAVE) {
                btSet.mode = CommScannerBtSettings.Mode.MASTER;
                connectedCommScanner.setBtSettings(btSet);
            } else if (Objects.equals(connectMode, "AUTO") && btSet.mode != CommScannerBtSettings.Mode.AUTO) {
                btSet.mode = CommScannerBtSettings.Mode.AUTO;
                connectedCommScanner.setBtSettings(btSet);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return btSet;
    }
}
